Firestore/Source/Remote/FSTSerializerBeta.mm

```diff
@@ -938,43 +936,45 @@ - (GCFSStructuredQuery_Filter *_Nullable)encodedFilters:(const Query::FilterList
   return result;
 }
 
-- (GCFSStructuredQuery_Filter *)encodedRelationFilter:(const RelationFilter &)filter {
+- (GCFSStructuredQuery_Filter *)encodedUnaryOrFieldFilter:(const FieldFilter &)filter {
   GCFSStructuredQuery_Filter *proto = [GCFSStructuredQuery_Filter message];
+
+  if (filter.op() == Operator::Equal) {
+    if (filter.value().is_null()) {
+      proto.unaryFilter.field = [self encodedFieldPath:filter.field()];
+      proto.unaryFilter.op = GCFSStructuredQuery_UnaryFilter_Operator_IsNull;
+      return proto;
+    }
+
+    if (filter.value().is_nan()) {
+      proto.unaryFilter.field = [self encodedFieldPath:filter.field()];
+      proto.unaryFilter.op = GCFSStructuredQuery_UnaryFilter_Operator_IsNan;
+      return proto;
+    }
+  }
+
   GCFSStructuredQuery_FieldFilter *fieldFilter = proto.fieldFilter;
   fieldFilter.field = [self encodedFieldPath:filter.field()];
-  fieldFilter.op = [self encodedRelationFilterOperator:filter.op()];
+  fieldFilter.op = [self encodedFieldFilterOperator:filter.op()];
   fieldFilter.value = [self encodedFieldValue:filter.value()];
   return proto;
 }
 
-- (std::shared_ptr<RelationFilter>)decodedRelationFilter:(GCFSStructuredQuery_FieldFilter *)proto {
+- (std::shared_ptr<FieldFilter>)decodedFieldFilter:(GCFSStructuredQuery_FieldFilter *)proto {
   FieldPath fieldPath = FieldPath::FromServerFormat(util::MakeString(proto.field.fieldPath));
-  Filter::Operator op = [self decodedRelationFilterOperator:proto.op];
+  Filter::Operator op = [self decodedFieldFilterOperator:proto.op];
   FieldValue value = [self decodedFieldValue:proto.value];
-  return std::make_shared<RelationFilter>(std::move(fieldPath), op, std::move(value));
-}
-
-- (GCFSStructuredQuery_Filter *)encodedUnaryFilter:(const Filter &)filter {
-  GCFSStructuredQuery_Filter *proto = [GCFSStructuredQuery_Filter message];
-  proto.unaryFilter.field = [self encodedFieldPath:filter.field()];
-  if (filter.type() == Filter::Type::kNanFilter) {
-    proto.unaryFilter.op = GCFSStructuredQuery_UnaryFilter_Operator_IsNan;
-  } else if (filter.type() == Filter::Type::kNullFilter) {
-    proto.unaryFilter.op = GCFSStructuredQuery_UnaryFilter_Operator_IsNull;
-  } else {
-    HARD_FAIL("Unrecognized filter: %s", filter.ToString());
-  }
-  return proto;
+  return FieldFilter::Create(std::move(fieldPath), op, std::move(value));
 }
 
 - (std::shared_ptr<Filter>)decodedUnaryFilter:(GCFSStructuredQuery_UnaryFilter *)proto {
   FieldPath field = FieldPath::FromServerFormat(util::MakeString(proto.field.fieldPath));
   switch (proto.op) {
     case GCFSStructuredQuery_UnaryFilter_Operator_IsNan:
```

Comment:
Ultranit: in `encodedUnaryOrFieldFilter`, the order is reverse (first null, then NaN); can this be consistent?

Reply:
Fixed.